src/TableOfContents/index.tsx

```diff
@@ -107,13 +109,26 @@ function TableOfContentsInner<T extends TableOfContentsItem = TableOfContentsIte
 
   // an array of functions. Invoking the N-th function toggles the expanded flag on the N-th content item
   const toggleExpandedFunctions = React.useMemo(() => {
-    return range(contents.length).map(i => () =>
-      setExpanded(current => ({
-        ...current,
-        [i]: !current[i],
-      })),
-    );
-  }, [contents.length]);
+    return range(contents.length).map(i => () => {
+      let childrenToCollapse = {};
+      if (expanded[i]) {
+        const item = contents[i];
+        const children = findDescendantIndices(item.depth ?? 0, i, contents.slice(i + 1));
+        childrenToCollapse = children.reduce((obj, index) => {
+          obj[index] = false;
+          return obj;
+        }, {});
+      }
+      setExpanded(current => {
+        return {
+          ...current,
+          [i]: !current[i],
+          ...childrenToCollapse,
+        };
+      });
```

Comment:
There was some reason I moved some of the logic to inside the callback function. Likely it's safer there, but in any case, I'd rather have it at a single place. What do you think of
```tsx
      setExpanded(current => {
        let childrenToCollapse = {};
        if (current[i]) {
          const item = contents[i];
          const children = findDescendantIndices(item.depth ?? 0, i, contents.slice(i + 1));
          childrenToCollapse = children.reduce((obj, index) => {
            obj[index] = false;
            return obj;
          }, {});
        }

        return {
          ...current,
          [i]: !current[i],
          ...childrenToCollapse,
        };
      });
```

And while you're there you could consider
```tsx
childrenToCollapse = Object.fromEntries(children.map(i => [i, false]));
```
which is the same as your `reduce`.

Reply:
Yup, that would be good, I've changed it. 
`fromEntries` indeed might be better - more readable too.